File: src/1377-Frog-Position-After-T-Seconds/1377.js
var frogPosition = function(n, edges, t, target) {
    if (n == 1) return 1.0;
    
    let g = [];
    for (let i = 0; i <= n; i++) g[i] = [];
    for (let it of edges) {
        g[it[0]].push(it[1]);
        g[it[1]].push(it[0]);
    }
    
    let dfs = function(fa, cur, tt) {
        if (cur != 1 && g[cur].length == 1 || tt == 0) {
            return cur == target;
        }
        
        let res = 0;
        for (let i of g[cur]) {
            if (i != fa) res = Math.max(res, dfs(cur, i, tt - 1));
        }
        return res / (g[cur].length - (cur != 1));
    }
    
    return dfs(-1, 1, t);
};

File: src/1377-Frog-Position-After-T-Seconds/1377.go
var g [][]int

func frogPosition(n int, edges [][]int, t int, target int) float64 {
    if n == 1 {
        return 1.0
    }
    
    g = make([][]int, n + 1)
    for i := 0; i <= n; i++ {
        g[i] = []int{}
    }
    for _, it := range edges {
        g[it[0]] = append(g[it[0]], it[1])
        g[it[1]] = append(g[it[1]], it[0])
    }
    
    return dfs(-1, 1, t, target)
}

func dfs(fa, cur, t, target int) float64 {
    if cur != 1 && len(g[cur]) == 1 || t == 0 {
        if cur == target {
            return 1
        }
        return 0
    }
    
    var res float64 = 0
    for _, i := range g[cur] {
        if i != fa {
            res = math.Max(res, dfs(cur, i, t - 1, target))
        }
    }
    if cur != 1 {
        return res / float64(len(g[cur]) - 1)
    }
    return res / float64(len(g[cur]))
}